File: bindings/python/tests/corrupt_fixture.py
"""Runtime corrupt-SSTable fixture generation for the abort-safety harness.

Issue #1437. No large binaries are committed to the repo. Given a temp dir,
this module copies the real ``test_basic.simple_table`` SSTable directory and
mutates ``nb-1-big-Data.db`` in one of two modes, both of which the harness
exercises:

- ``truncate``: copy the SSTable dir, then truncate ``nb-1-big-Data.db`` to
  roughly 50% of its byte length.
- ``bitflip``: copy the whole SSTable dir, then XOR ``0x01`` into a byte in the
  middle of ``nb-1-big-Data.db``.

The sibling ``-Statistics.db`` / ``-Index.db`` / ``-Summary.db`` /
``-CompressionInfo.db`` / ``-TOC.txt`` components are kept intact so that
``cqlite.open()`` proceeds far enough to actually read the corrupt Data.db
during a scan (that is where the corrupt-input panic paths live).

Dataset rule (issue #1437): if the real source Data.db is present but empty or
unreadable, callers must FAIL LOUDLY — never skip. ``ensure_source_readable``
raises so the pytest layer can turn that into a hard failure.
"""

from __future__ import annotations

import shutil
from pathlib import Path

KEYSPACE = "test_basic"
TABLE = "simple_table"
DATA_COMPONENT = "nb-1-big-Data.db"
COMPRESSION_COMPONENT = "nb-1-big-CompressionInfo.db"
TOC_COMPONENT = "nb-1-big-TOC.txt"
MODES = ("truncate", "bitflip")

# Do not copy the multi-megabyte derived JSONL golden; the reader only consults
# the TOC-listed binary components. The small ``-TOC.txt`` MUST be kept (the
# issue requires it), so we scope the ignore to the JSONL only.
_IGNORE = shutil.ignore_patterns("*.jsonl")


def source_table_dir(datasets_root: Path) -> Path | None:
    """Return the real ``simple_table-<uuid>`` dir under ``datasets_root``.

    ``datasets_root`` is the ``sstables/`` directory (the conftest ``DATASETS``
    constant). Returns ``None`` when no matching table dir with a Data.db is
    present (dataset simply not fetched).
    """
    ks_dir = Path(datasets_root) / KEYSPACE
    if not ks_dir.is_dir():
        return None
    for candidate in sorted(ks_dir.glob(f"{TABLE}-*")):
        if (candidate / DATA_COMPONENT).is_file():
            return candidate
    return None


def ensure_source_readable(datasets_root: Path) -> Path:
    """Return the source table dir or raise loudly when it is broken.

    Raises ``FileNotFoundError`` when the dataset is absent (caller decides
    skip-vs-fail) and ``ValueError`` when the Data.db is present but empty —
    which is a corrupt fixture source, never a skip (issue #1437).
    """
    src = source_table_dir(datasets_root)
    if src is None:
        raise FileNotFoundError(
            f"No {KEYSPACE}.{TABLE} SSTable with {DATA_COMPONENT} under {datasets_root}"
        )
    data = src / DATA_COMPONENT
    if data.stat().st_size == 0:
        raise ValueError(f"Source {data} is present but empty (corrupt dataset)")
    return src


def _drop_compression_info(dest_table: Path) -> None:
    """Remove ``CompressionInfo.db`` (and its TOC line) from a copied table.

    This forces the reader down the *uncompressed* Data.db path, which parses
    raw VInt/row lengths directly. That is where the corrupt-input panics live
    (ZigZag lengths etc.); with the compression sidecar present, the Snappy
    decompression layer contains any corruption and the panics are never
    reached. A missing/absent CompressionInfo.db is itself a plausible
    real-world corruption, so this is a valid abort-safety scenario — not a
    contrivance.
    """
    comp = dest_table / COMPRESSION_COMPONENT
    if comp.exists():
        comp.unlink()
    toc = dest_table / TOC_COMPONENT
    if toc.exists():
        kept = [
            line
            for line in toc.read_text().splitlines()
            if "CompressionInfo" not in line
        ]
        toc.write_text("\n".join(kept) + "\n")


def make_corrupt_fixture(
    dest_parent: Path,
    datasets_root: Path,
    mode: str,
    *,
    expose_uncompressed: bool = False,
) -> Path:
    """Build a corrupt copy of the SSTable under ``dest_parent`` and return the
    ``sstables/`` root to pass to :func:`cqlite.open`.

    ``mode`` is one of :data:`MODES`. When ``expose_uncompressed`` is True the
    copy also drops ``CompressionInfo.db`` so the corrupt Data.db is read on the
    raw (uncompressed) parse path — see :func:`_drop_compression_info`.
    """
    if mode not in MODES:
        raise ValueError(f"unknown mode {mode!r}; expected one of {MODES}")

    src = ensure_source_readable(datasets_root)
    dest_root = Path(dest_parent) / "sstables"
    dest_table = dest_root / KEYSPACE / src.name
    shutil.copytree(src, dest_table, ignore=_IGNORE)

    if expose_uncompressed:
        _drop_compression_info(dest_table)

    data = dest_table / DATA_COMPONENT
    length = data.stat().st_size
    if mode == "truncate":
        with open(data, "r+b") as fh:
            fh.truncate(length // 2)
    else:  # bitflip
        offset = length // 2
        with open(data, "r+b") as fh:
            fh.seek(offset)
            (byte,) = fh.read(1)
            fh.seek(offset)
            fh.write(bytes([byte ^ 0x01]))

    return dest_root

File: bindings/node/__test__/corrupt-fixture.js
/**
 * Runtime corrupt-SSTable fixture generation for the abort-safety harness.
 *
 * Issue #1437. No large binaries are committed. Given a temp dir this copies
 * the real `test_basic.simple_table` SSTable directory and mutates
 * `nb-1-big-Data.db` in one of two modes, both exercised by the harness:
 *   - "truncate": copy the dir, then truncate Data.db to ~50% of its length.
 *   - "bitflip":  copy the whole dir, then XOR 0x01 into a middle byte.
 *
 * Sibling components (-Statistics.db / -Index.db / -Summary.db /
 * -CompressionInfo.db / -TOC.txt) are kept so Database.open() proceeds far
 * enough to read the corrupt Data.db during a scan.
 *
 * Dataset rule (issue #1437): if the real source Data.db is missing or
 * zero-length, THROW (never silently `return`); a broken source is a hard
 * failure, not a skippable condition.
 */
'use strict';

const fs = require('fs');
const path = require('path');
// The canonical table-directory predicates, shared with the golden lookups so all three
// consumers and the manifest agree on what a table directory IS (#3493).
const { isTableDirFor, isCommittedTableDir } = require('./parity-utils.js');

const KEYSPACE = 'test_basic';
const TABLE = 'simple_table';
const DATA_COMPONENT = 'nb-1-big-Data.db';
const COMPRESSION_COMPONENT = 'nb-1-big-CompressionInfo.db';
const TOC_COMPONENT = 'nb-1-big-TOC.txt';
const MODES = ['truncate', 'bitflip'];

/**
 * Classify the source table directory: `ok` | `broken` | `absent`.
 *
 * THE THREE-VALUED ANSWER IS THE POINT (roborev #3493 round 52). `sourceTableDir` returns a
 * directory or `null`, and round 48 added a nonempty-regular-file filter to it -- which
 * silently converted "present but UNUSABLE" into "ABSENT". `abort-safety.test.js` keys its
 * gating on exactly that distinction: `broken` is a HARD FAILURE, while a non-strict
 * `absent` is a real `test.skip`. So a truncated fixture stopped hard-failing and started
 * SKIPPING, inverting #1437's stated design ("a broken source is a hard failure, not a
 * skippable condition") and making that test's `broken` branch unreachable.
 *
 * `broken` means a Data.db ENTRY EXISTS in a canonical candidate but is not a nonempty
 * regular file -- zero-length, a directory, a dangling symlink. A canonical directory with
 * no Data.db entry at all stays `absent`, which is the pre-round-48 behaviour and the
 * honest reading: the fixture was never fetched, as opposed to fetched and damaged.
 *
 * A VALID candidate still WINS over a broken sibling. That is deliberately laxer than
 * `check-dataset-manifest.sh`, which disqualifies a table when ANY candidate is unusable --
 * and the two are right for different reasons: Jest's discovery picks a directory blind, so
 * the manifest must assume the worst, while this function picks deterministically and can
 * simply choose the good one.
 *
 * @param {string} sstablesRoot
 * @returns {{status: 'ok', dir: string} | {status: 'broken', dir: string, reason: string}
 *           | {status: 'absent', reason: string}}
 */
function classifyTableDir(sstablesRoot) {
  const ksDir = path.join(sstablesRoot, KEYSPACE);
  if (!fs.existsSync(ksDir) || !fs.statSync(ksDir).isDirectory()) {
    return { status: 'absent', reason: `No ${KEYSPACE} keyspace directory under ${sstablesRoot} (issue #1437)` };
  }
  const candidates = fs
    .readdirSync(ksDir, { withFileTypes: true })
    // A REAL directory. `Dirent.isDirectory()` is FALSE for a symlink, which is exactly
    // what Jest's discovery and the manifest's `[ -L "$cand" ] && continue` both do, so a
    // symlinked table dir is invisible to them and must be invisible here too.
    //
    // Not merely a consistency point -- it is what stops this harness DESTROYING THE
    // SOURCE CORPUS (round 49). `fs.cpSync` preserves a symlink by default, so a symlinked
    // table dir was copied AS A SYMLINK and the truncate/bitflip below then wrote THROUGH
    // it into the real fixture: truncating the "copy" cut the original from 12 bytes to 4.
    // On a shared machine-local corpus that is damage to every other lane on the box.
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .filter((name) => isTableDirFor(name, TABLE) && isCommittedTableDir(KEYSPACE, name))
    .map((name) => path.join(ksDir, name))
    .sort();

  const usable = candidates.filter((dir) => isNonemptyFile(path.join(dir, DATA_COMPONENT)));
  if (usable.length > 0) {
    return { status: 'ok', dir: usable[0] };
  }
  const damaged = candidates.find((dir) => entryExists(path.join(dir, DATA_COMPONENT)));
  if (damaged !== undefined) {
    return {
      status: 'broken',
      dir: damaged,
      reason:
        `Source ${path.join(damaged, DATA_COMPONENT)} is present but not a nonempty ` +
        `regular file (zero-length, a directory, or a dangling symlink) (issue #1437)`,
    };
  }
  return {
    status: 'absent',
    reason: `No ${KEYSPACE}.${TABLE} SSTable with ${DATA_COMPONENT} under ${sstablesRoot} (issue #1437)`,
  };
}

/**
 * The usable source table dir, or null. Thin wrapper over `classifyTableDir` -- callers
 * that need to tell `broken` from `absent` must use the classifier (see its header).
 *
 * @param {string} sstablesRoot
 * @returns {string|null}
 */
function sourceTableDir(sstablesRoot) {
  const c = classifyTableDir(sstablesRoot);
  return c.status === 'ok' ? c.dir : null;
}

/**
 * True iff a directory ENTRY exists at `p`, whatever its type.
 *
 * `lstatSync`, not `existsSync`: the latter follows symlinks and so answers FALSE for a
 * DANGLING one -- which would classify a dangling-symlink Data.db as `absent` when it is
 * exactly the damaged-fixture case `broken` exists to report.
 *
 * @param {string} p
 * @returns {boolean}
 */
function entryExists(p) {
  try {
    fs.lstatSync(p);
    return true;
  } catch (_e) {
    return false;
  }
}

/**
 * True iff `p` is a NONEMPTY REGULAR FILE (following symlinks).
 *
 * `fs.existsSync` is true for a directory and false for a dangling symlink, and says
 * nothing about size -- and a zero-length Data.db is exactly what a truncated fetch
 * leaves, which this harness would then "corrupt" and assert against.
 *
 * @param {string} p
 * @returns {boolean}
 */
function isNonemptyFile(p) {
  try {
    const st = fs.statSync(p);
    return st.isFile() && st.size > 0;
  } catch (_e) {
    return false;
  }
}

/**
 * Force the reader down the uncompressed Data.db path by dropping the
 * CompressionInfo sidecar (and its TOC line). The raw VInt/row parser is where
 * the audited corrupt-input panics live; with compression present, Snappy
 * decompression contains any corruption before the panics are reached. A
 * missing CompressionInfo.db is itself a plausible real-world corruption.
 * @param {string} destTable
 */
function dropCompressionInfo(destTable) {
  const comp = path.join(destTable, COMPRESSION_COMPONENT);
  if (fs.existsSync(comp)) {
    fs.rmSync(comp);
  }
  const toc = path.join(destTable, TOC_COMPONENT);
  if (fs.existsSync(toc)) {
    const kept = fs
      .readFileSync(toc, 'utf8')
      .split('\n')
      .filter((line) => !line.includes('CompressionInfo'));
    fs.writeFileSync(toc, kept.join('\n') + '\n');
  }
}

/**
 * Build a corrupt copy under `destParent` and return the `sstables/` root to
 * pass to Database.open. Throws when the source Data.db is missing/empty.
 * @param {string} destParent
 * @param {string} sstablesRoot
 * @param {'truncate'|'bitflip'} mode
 * @param {{exposeUncompressed?: boolean}} [opts] - when exposeUncompressed is
 *   true, also drop CompressionInfo.db to read Data.db on the raw parse path.
 * @returns {string}
 */
function makeCorruptFixture(destParent, sstablesRoot, mode, opts = {}) {
  if (!MODES.includes(mode)) {
    throw new Error(`unknown mode ${mode}; expected one of ${MODES.join(', ')}`);
  }
  const src = sourceTableDir(sstablesRoot);
  if (src === null) {
    throw new Error(
      `No ${KEYSPACE}.${TABLE} SSTable with ${DATA_COMPONENT} under ${sstablesRoot}`
    );
  }
  const srcData = path.join(src, DATA_COMPONENT);
  if (fs.statSync(srcData).size === 0) {
    throw new Error(`Source ${srcData} present but empty (issue #1437)`);
  }

  const destRoot = path.join(destParent, 'sstables');
  const destTable = path.join(destRoot, KEYSPACE, path.basename(src));
  fs.cpSync(src, destTable, {
    recursive: true,
    // COPY WHAT A SYMLINK POINTS AT, never the link (#3493 round 49). Defence in depth
    // behind sourceTableDir's symlink rejection: that guard covers the TABLE DIRECTORY,
    // while a symlinked COMPONENT inside a real directory would still be copied as a link
    // and the mutation below would write through it into the source corpus.
    dereference: true,
    // Skip only the multi-megabyte JSONL golden; the small -TOC.txt MUST be
    // kept (issue #1437). The reader consults only TOC-listed binary files.
    filter: (s) => !s.endsWith('.jsonl'),
  });

  if (opts.exposeUncompressed) {
    dropCompressionInfo(destTable);
  }

  const destData = path.join(destTable, DATA_COMPONENT);
  const length = fs.statSync(destData).size;
  if (mode === 'truncate') {
    fs.truncateSync(destData, Math.floor(length / 2));
  } else {
    const offset = Math.floor(length / 2);
    const fd = fs.openSync(destData, 'r+');
    try {
      const buf = Buffer.alloc(1);
      fs.readSync(fd, buf, 0, 1, offset);
      buf[0] ^= 0x01;
      fs.writeSync(fd, buf, 0, 1, offset);
    } finally {
      fs.closeSync(fd);
    }
  }
  return destRoot;
}

module.exports = {
  KEYSPACE,
  TABLE,
  DATA_COMPONENT,
  MODES,
  classifyTableDir,
  sourceTableDir,
  makeCorruptFixture,
};
